pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PinotSegmentUploadRestletResource.java

```diff
@@ -328,9 +346,26 @@ private SuccessResponse uploadSegmentInternal(FormDataMultiPart multiPart, Strin
             throw new ControllerApplicationException(LOGGER,
                 "Failed to get segment fetcher for download URI: " + downloadURI, Response.Status.BAD_REQUEST);
           }
-
-          // Download segment tar file to local
-          segmentFetcher.fetchSegmentToLocal(downloadURI, tempTarredSegmentFile);
+          if (segmentMetadata == null) {
+            // Download segment tar file to local
+            segmentFetcher.fetchSegmentToLocal(downloadURI, tempTarredSegmentFile);
+          } else {
+            // Read segment metadata file and directly use that information to update zk
```

Comment:
How do we guard against a client bug (or worse, malicious client) that may send wrong metadata? (a design question, I suppose)

Reply:
We decided during design review that we wouldn't take care of this case. Have some notes in the design document about this (basically the problem is difficult because without downloading segment, can always be injected)